R/proxy.R

```diff
@@ -1,28 +1,40 @@
-#' Extract underlying data
+#' Proxy and restore
 #'
 #' @description
 #'
-#' Extract the data underlying an S3 vector object, i.e. the underlying
-#' (named) atomic vector or list.
+#' `vec_proxy()` returns the memory representation of a vector. You
+#' normally don't need to implement it, except in these cases:
 #'
-#' * `vec_data()` returns unstructured data. The only attributes
-#'   preserved are names, dims, and dimnames.
+#' - S3 lists are considered scalars by default. This is the safe
+#'   choice for list objects such as returned by `stats::lm()`. To
+#'   declare that your S3 list class is a vector, give it an identity
+#'   proxy (i.e. a proxy method that just returns its input).
 #'
-#'   Currently, due to the underlying memory architecture of R, this
-#'   creates a full copy of the data.
+#' - Some vector types have vectorised attributes, i.e. metadata for
+#'   each element of the vector. These _record types_ are implemented
+#'   in vctrs by returning a data frame in the proxy method.
 #'
-#' * `vec_proxy()` may return structured data. This generic is the
-#'   main customisation point in vctrs, along with [vec_restore()].
-#'   See the section below to learn when you should implement
-#'   `vec_proxy()`.
+#' - When you're implementing a vector on top of a non-vector type,
+#'   like an environment or an S4 object. This is currently only
+#'   partially supported.
 #'
-#'   Methods must return a vector type. Records and data frames will
-#'   be processed rowwise.
+#' `vec_restore()` is the inverse operation of `vec_proxy()`. It
+#' should only be called on vector proxies.
 #'
-#' @param x A vector or object implementing `vec_proxy()`.
-#' @return The data underlying `x`, free from any attributes except the names.
+#' - It undoes the transformations of `vec_proxy()`.
+#'
+#' - It restores attributes and classes. These may be lost when the
+#'   memory values are manipulated. For example slicing a subset of a
+#'   vector's proxy causes a new proxy to be allocated.
+#'
+#' By default vctrs restores all attributes and classes
+#' automatically. You only need to implement a `vec_restore()` method
+#' if your class has attributes that depend on the data.
+#'
+#' @param x A vector.
+#' @inheritParams ellipsis::dots_empty
```

Comment:
I think this should be `@keywords` internal and mention that this mechanism is still under development.

Reply:
Now marked as experimental.